pkg/components/pubsub/pluggable_test.go

```diff
@@ -0,0 +1,293 @@
+	}
+	return &proto.PubSubInitResponse{}, s.initErr
+}
+
+func (s *server) Features(context.Context, *proto.FeaturesRequest) (*proto.FeaturesResponse, error) {
+	s.featuresCalled.Add(1)
+	return &proto.FeaturesResponse{}, s.featuresErr
+}
+
+func (s *server) Publish(_ context.Context, req *proto.PublishRequest) (*proto.PublishResponse, error) {
+	s.publishCalled.Add(1)
+	if s.onPublishCalled != nil {
+		s.onPublishCalled(req)
+	}
+	return &proto.PublishResponse{}, s.publishErr
+}
+
+func (s *server) Ping(context.Context, *proto.PingRequest) (*proto.PingResponse, error) {
+	s.pingCalled.Add(1)
+	return &proto.PingResponse{}, s.pingErr
+}
+
+func TestPubSubPluggableCalls(t *testing.T) {
+	getPubSub := testingGrpc.TestServerFor(testLogger, func(s *grpc.Server, svc *server) {
+		proto.RegisterPubSubServer(s, svc)
+	}, func(cci grpc.ClientConnInterface) *grpcPubSub {
+		client := proto.NewPubSubClient(cci)
+		pubsub := fromConnector(testLogger, pluggable.NewGRPCConnector(components.Pluggable{}, proto.NewPubSubClient))
+		pubsub.Client = client
+		return pubsub
+	})
+
+	if runtime.GOOS != "windows" {
+		t.Run("test init should populate features and call grpc init", func(t *testing.T) {
+			const (
+				fakeName          = "name"
+				fakeType          = "type"
+				fakeVersion       = "v1"
+				fakeComponentName = "component"
+				fakeSocketFolder  = "/tmp"
+			)
+
+			uniqueID := guuid.New().String()
+			socket := fmt.Sprintf("%s/%s.sock", fakeSocketFolder, uniqueID)
+			defer os.Remove(socket)
+
+			connector := pluggable.NewGRPCConnectorWithFactory(func(string) string {
+				return socket
+			}, proto.NewPubSubClient)
+			defer connector.Close()
+
+			listener, err := net.Listen("unix", socket)
+			require.NoError(t, err)
+			defer listener.Close()
+			s := grpc.NewServer()
+			srv := &server{}
+			proto.RegisterPubSubServer(s, srv)
+			go func() {
+				if serveErr := s.Serve(listener); serveErr != nil {
+					testLogger.Debugf("Server exited with error: %v", serveErr)
+				}
+			}()
+
+			ps := fromConnector(testLogger, connector)
+			err = ps.Init(pubsub.Metadata{
+				Base: contribMetadata.Base{},
+			})
+
+			require.NoError(t, err)
+			assert.Equal(t, int64(1), srv.featuresCalled.Load())
+			assert.Equal(t, int64(1), srv.initCalled.Load())
+		})
+	}
+
+	t.Run("features should return the component features'", func(t *testing.T) {
+		ps, cleanup, err := getPubSub(&server{})
+		require.NoError(t, err)
+		defer cleanup()
+		assert.Empty(t, ps.Features())
+		ps.features = []pubsub.Feature{pubsub.FeatureMessageTTL}
+		assert.NotEmpty(t, ps.Features())
+		assert.Equal(t, ps.Features()[0], pubsub.FeatureMessageTTL)
+	})
+
+	t.Run("publish should call publish grpc method", func(t *testing.T) {
+		const fakeTopic = "fakeTopic"
+
+		svc := &server{
+			onPublishCalled: func(req *proto.PublishRequest) {
+				assert.Equal(t, req.Topic, fakeTopic)
+			},
+		}
+		ps, cleanup, err := getPubSub(svc)
+		require.NoError(t, err)
+		defer cleanup()
```

Comment:
If `err` no cleanup is required, correct?

Reply:
Yes. This is the expected behavior when returning error. If the function has created unmanaged resources so it should dispose then given that an error is returned and the cleanup function should be `nil` in such cases